airflow/contrib/hooks/gcf_hook.py

```diff
@@ -0,0 +1,107 @@
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+from time import sleep
+
+import requests
+from googleapiclient.discovery import build
+
+from airflow.contrib.hooks.gcp_api_base_hook import GoogleCloudBaseHook
+
+NUM_RETRIES = 5
+
+
+class GCFHook(GoogleCloudBaseHook):
+    """Hook for Google Cloud Functions APIs."""
+    MAX_RETRIES = 1000
+
+    def __init__(self,
+                 api_version,
+                 gcp_conn_id='google_cloud_default',
+                 delegate_to=None):
+        super(GCFHook, self).__init__(gcp_conn_id, delegate_to)
+        self.api_version = api_version
+
+    def get_conn(self):
+        """Returns a Google Cloud Functions service object."""
+        http_authorized = self._authorize()
+        return build(
+            'cloudfunctions', self.api_version, http=http_authorized,
+            cache_discovery=False)
+
+    def list_functions(self, location):
+        list_response = self.get_conn().projects().locations().functions().list(
+            parent=location).execute(num_retries=NUM_RETRIES)
+        return list_response.get("functions", [])
+
+    def create_new_function(self, location, body):
+        response = self.get_conn().projects().locations().functions().create(
+            location=location,
+            body=body
+        ).execute(num_retries=NUM_RETRIES)
+        operation_name = response["name"]
+        return self._wait_for_operation_to_complete(operation_name)
+
+    def update_function(self, name, body, fields):
+        response = self.get_conn().projects().locations().functions().patch(
+            updateMask=",".join(fields),
+            name=name,
+            body=body
+        ).execute(num_retries=NUM_RETRIES)
+        operation_name = response["name"]
+        return self._wait_for_operation_to_complete(operation_name)
+
+    def upload_function_zip(self, parent, zip_path):
+        response = self.get_conn().projects().locations().functions().generateUploadUrl(
+            parent=parent
+        ).execute(num_retries=NUM_RETRIES)
+        uploadURL = response.get('uploadUrl')
+        with open(zip_path, 'rb') as fp:
+            requests.put(
+                uploadURL,
+                data=fp.read(),
+                headers={
+                    'Content-type': 'application/zip',
+                    'x-goog-content-length-range': '0,104857600',
+                }
+            )
+        return uploadURL
+
+    def delete_function(self, name):
+        response = self.get_conn().projects().locations().functions().delete(
+            name=name).execute(num_retries=NUM_RETRIES)
+        operation_name = response["name"]
+        return self._wait_for_operation_to_complete(operation_name)
+
+    def _wait_for_operation_to_complete(self, operation_name):
+        service = self.get_conn()
+        retries = 0
+        while True:
+            if retries > self.MAX_RETRIES:
```

Comment:
I think we can infinitely poll and remove this MAX_RETRIES functionality for 2 reasons.

1. MAX_RETRIES is a little non-intuitive, as users would ideally want to say a max timeout
2. Airflow operators have a timeout argument to perform this functionality.

Reply:
Done by @sprzedwojski